File: webhooks/typescript/verify.ts
// Verifies the X-Fletch-Signature header on a webhook delivery from Fletch.
// The header is `t=<unix seconds>,v1=<hex>` where the hex is HMAC-SHA256 over
// the exact string "<t>.<raw body>" keyed with the endpoint's whsec_ secret.
// The timestamp is inside the signed string, so a captured body cannot be
// replayed later with a fresh header; `toleranceSeconds` bounds how old (or
// how far in the future) a delivery may be. Node only: node:crypto.

import { createHmac, timingSafeEqual } from "node:crypto";

export interface VerifyOptions {
  // Seconds either side of `now` that a timestamp is accepted; default 300.
  toleranceSeconds?: number;
  // Unix seconds to judge the timestamp against; default the clock.
  now?: number;
}

export interface ParsedSignature {
  timestamp: number;
  signature: string;
}

const SIGNATURE_HEX_LENGTH = 64;

// Unknown fields are ignored so a later scheme (v2=…) can be added beside v1
// without breaking receivers on this version.
export function parseSignatureHeader(header: string): ParsedSignature | null {
  let timestamp: string | null = null;
  let signature: string | null = null;
  for (const part of header.split(",")) {
    const separator = part.indexOf("=");
    if (separator < 0) {
      continue;
    }
    const key = part.slice(0, separator).trim();
    const value = part.slice(separator + 1).trim();
    if (key === "t") {
      timestamp = value;
    } else if (key === "v1") {
      signature = value;
    }
  }
  if (timestamp === null || signature === null) {
    return null;
  }
  if (!/^\d+$/.test(timestamp) || !/^[0-9a-fA-F]+$/.test(signature) || signature.length !== SIGNATURE_HEX_LENGTH) {
    return null;
  }
  return { timestamp: Number(timestamp), signature: signature.toLowerCase() };
}

export function signWebhookBody(secret: string, body: string | Uint8Array, timestampSeconds: number): string {
  const digest = createHmac("sha256", secret).update(`${timestampSeconds}.`).update(body).digest("hex");
  return `t=${timestampSeconds},v1=${digest}`;
}

// `body` must be the raw request bytes (or the exact string they decode to),
// never a re-serialized object: a reordered key or a changed space changes
// the digest.
export function verifyFletchSignature(secret: string, header: string, body: string | Uint8Array, options: VerifyOptions = {}): boolean {
  const parsed = parseSignatureHeader(header);
  if (!parsed) {
    return false;
  }
  const tolerance = options.toleranceSeconds ?? 300;
  const now = options.now ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - parsed.timestamp) > tolerance) {
    return false;
  }
  const expected = createHmac("sha256", secret).update(`${parsed.timestamp}.`).update(body).digest();
  const given = Buffer.from(parsed.signature, "hex");
  return given.length === expected.length && timingSafeEqual(given, expected);
}

File: webhooks/go/verify.go
// Package fletchwebhook verifies the X-Fletch-Signature header on a webhook
// delivery from Fletch.
//
// The header is "t=<unix seconds>,v1=<hex>" where the hex is HMAC-SHA256 over
// the exact bytes "<t>." + body keyed with the endpoint's whsec_ secret. The
// timestamp is inside the signed string, so a captured body cannot be replayed
// later with a fresh header; the tolerance bounds how old (or how far in the
// future) a delivery may be. Standard library only.
package fletchwebhook

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"strings"
	"time"
)

const signatureHexLength = 64

// ParsedSignature is the timestamp and digest carried by a well-formed header.
type ParsedSignature struct {
	Timestamp int64
	Signature []byte
}

// ParseSignatureHeader returns the timestamp and digest, or ok=false when the
// header is malformed. Unknown fields are ignored so a later scheme (v2=...)
// can be added beside v1 without breaking receivers on this version.
func ParseSignatureHeader(header string) (ParsedSignature, bool) {
	var timestamp, signature string
	var haveTimestamp, haveSignature bool
	for _, part := range strings.Split(header, ",") {
		key, value, found := strings.Cut(part, "=")
		if !found {
			continue
		}
		switch strings.TrimSpace(key) {
		case "t":
			timestamp, haveTimestamp = strings.TrimSpace(value), true
		case "v1":
			signature, haveSignature = strings.TrimSpace(value), true
		}
	}
	if !haveTimestamp || !haveSignature {
		return ParsedSignature{}, false
	}
	if timestamp == "" || strings.Trim(timestamp, "0123456789") != "" {
		return ParsedSignature{}, false
	}
	seconds, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return ParsedSignature{}, false
	}
	if len(signature) != signatureHexLength {
		return ParsedSignature{}, false
	}
	digest, err := hex.DecodeString(signature)
	if err != nil {
		return ParsedSignature{}, false
	}
	return ParsedSignature{Timestamp: seconds, Signature: digest}, true
}

func digest(secret string, timestamp int64, body []byte) []byte {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(strconv.FormatInt(timestamp, 10)))
	mac.Write([]byte("."))
	mac.Write(body)
	return mac.Sum(nil)
}

// SignWebhookBody produces the header value Fletch would send for this body.
func SignWebhookBody(secret string, body []byte, timestamp int64) string {
	return "t=" + strconv.FormatInt(timestamp, 10) + ",v1=" + hex.EncodeToString(digest(secret, timestamp, body))
}

// Verify reports whether header signs body with secret and its timestamp is
// within tolerance of now. Pass the raw request bytes, never a re-serialized
// struct: a reordered key or a changed space changes the digest.
func Verify(secret, header string, body []byte, tolerance time.Duration, now time.Time) bool {
	parsed, ok := ParseSignatureHeader(header)
	if !ok {
		return false
	}
	age := now.Unix() - parsed.Timestamp
	if age < 0 {
		age = -age
	}
	if age > int64(tolerance/time.Second) {
		return false
	}
	return hmac.Equal(parsed.Signature, digest(secret, parsed.Timestamp, body))
}

// VerifyNow is Verify against the clock with the default five-minute tolerance.
func VerifyNow(secret, header string, body []byte) bool {
	return Verify(secret, header, body, 5*time.Minute, time.Now())
}
